File: js/sample.js
const logipar = require('./logipar/Logipar.js');
const cats = require("./cats.js");

const readline = require('readline').createInterface({
	input: process.stdin,
	output: process.stdout
})

console.log("-- Welcome to the cat library --")
let s = ""
readline.question("Please enter an input string: ", (str) => {
	s = str
	readline.close()
		
	lp = new logipar.Logipar()
	lp.overwrite(logipar.Token.AND, "et")
	lp.caseSensitive = false
	lp.parse(s)


	function expandXOR(n) {
		if (n.token.type === logipar.Token.XOR) {
			l = n.f(n.left)
			r = n.f(n.right)
			return "((" + l + " AND NOT " + r + ") OR (NOT " + l + " AND " + r + "))"
		}
		return null
	}

	const flattened = lp.stringify(expandXOR)

	console.log("\nOkay, it looks like you're looking for:", flattened, "\n\n")


	function fancyFilter(row, value) {
		value = value.replace('"', '')
		if (!value.includes(":")) { 
			// This is a dumb one that just checks the values against every column
			for(let field in row) {
				if (("" + row[field]).toLowerCase().includes(value.toLowerCase()))
					return true
			}
		} else {
			// We're specifying a specific field we want to look in
			let chunks = value.split(':');
			const field = chunks.shift();
			let val = ':'.join(chunks)
			if (row.hasOwnProperty(field))
				if (("" + row[field]).toLowerCase().includes(val.lower()))
					return true
		}
		return false
	}

	const f = lp.filterFunction(fancyFilter)


	data = cats.filter(f)
	if (data.length == 0)
		console.log("No matching entries found.")
		
	for(let i=0; i < data.length; i++)
		console.log(data[i]['Breed'])

	console.log("\nFound", data.length, "entries.")	
	
})

File: python/sample.py
import logipar
import json, os


with open(os.path.dirname(os.path.realpath(__file__)) + os.sep + 'cats.json') as json_file:
	data = json.load(json_file)

print("-- Welcome to the cat library --")
s = input("Please enter an input string: ")

l = logipar.Logipar()
l.overwrite(logipar.Token.AND, "et")
l.caseSensitive = False
l.parse(s)


def expandXOR(n):
	if n.token.type == logipar.Token.XOR:
		l = n.f(n.left)
		r = n.f(n.right)
		return "(({} AND NOT {}) OR (NOT {} AND {}))".format(l, r, l, r)
	return None

flattened = l.stringify(expandXOR)

print("\nOkay, it looks like you're looking for: {}\n\n".format(flattened))


def fancyFilter(row, value):
	value = value.replace('"', '')
	if ":" not in value: 
		# This is a dumb one that just checks the values against every column
		for field in row:
			if value.lower() in str(row[field]).lower():
				return True
	else:
		# We're specifying a specific field we want to look in
		chunks = value.split(':');
		field = chunks[0];
		val = ':'.join(chunks[1:])
		if field in row:
			if val.lower() in str(row[field]).lower():
				return True
	return False

f = l.filterFunction(fancyFilter)

data = list(filter(f, data))
if len(data) == 0:
	print("No matching entries found.")
	
for d in data:
	print("{}".format(d['Breed']))

print("\nFound {} entries.".format(len(data)))